### ISSUE TITLE
Block 13: fused MoE MMQ breaks MTP verify on Qwen3.6-35B-A3B UD-Q4_K_XL (draft acceptance 0/242)

### ISSUE THREAD

**Issue by laurentlp** [stew675/llama-cpp-rdna-boosts#12]
## Summary

Testing the 13-block set on a single R9700 (gfx1201). On Qwen3.6-35B-A3B UD-Q4_K_XL with embedded MTP heads, MTP draft acceptance collapses to **0/242** with the block-13 MoE fusion active, and fully recovers to **41/87 (0.47)** with it disabled via `GGML_CUDA_DISABLE_MOE_MMQ_FUSION=1`. Same binary, same flags, same model — bisected to the one env knob. This looks like a third instance of the 2026-09-02 MoE verify-divergence class that the rms-norm-fold gate doesn't cover.

## Build

- llama.cpp @ `9cffdcc80` + blocks 01–13 via `scripts/apply-all.sh` (clean apply), server reports `0.3.0-dev (build 10786, commit b223d512c)`
- `cmake -DGGML_HIP=ON -DGGML_HIP_RCCL=1 -DGGML_HIP_GRAPHS=ON -DGPU_TARGETS="gfx1201"` (no NATIVE), ROCm 7.2 builder, single GPU
- Only difference between the two runs below is the env knob (same container image)

## Model / flags

- `Qwen3.6-35B-A3B-UD-Q4_K_XL.gguf` + `mmproj-Qwen3.6-35B-A3B-BF16.gguf` (Unsloth UD)
- `--n-gpu-layers 999 --flash-attn on --cache-type-k q8_0 --cache-type-v q8_0 --ctx-size 32768 --spec-type draft-mtp --spec-draft-n-max 4`, temp 0

## Numbers (same 64-token temp-0 probe + pp/tg bench)

| run | draft accepted | tg (64tok probe) | tg512 bench | pp19k bench |
|---|---|---|---|---|
| fusion ON (default) | 0/242 (0.0) | 38.8 t/s | 39.4 t/s | ~2950 t/s |
| `GGML_CUDA_DISABLE_MOE_MMQ_FUSION=1` | 41/87 (0.47) | 83.2 t/s | 125.1 t/s | ~2857 t/s |
| stock Vulkan ref (same model/flags) | n/a | n/a | 126.8 t/s | 3184.5 t/s |

With fusion off, decode reaches full parity with stock Vulkan (125.1 vs 126.8) while prefill only drops ~3% — so the fusion buys almost nothing here and destroys MTP.

## Interpretation

The fused MoE gate+up+GLU MMQ path appears to corrupt multi-token MTP verify logits on this quant/model (Q4_K_XL MoE), exactly the failure mode of the 2026-09-02 items — but the current gate (rms-norm fold restricted to single-token MMID) doesn't catch it. Possibly the fusion half needs the same single-vs-multi-token gating, or the fused prefill kernel writes something the verify path misreads for this quant.

Happy to run any diagnostic build/probe on this box (single R9700, local CI runner, ~20 min rebuilds) — e.g. coherence dumps, per-block isolation, other quants of the same model if useful.

Thanks for the set — blocks 01–11 are behaving well on my dense Q6_K model (acceptance 0.61–0.65, small decode gap only), it's specifically the MoE fusion + MTP verify combo that falls over here.

**Comment by laurentlp** [comment]
Update before you spend time on this: the headline number **failed to reproduce**. A second identical load of the same `nofusion` entry did tg512 @ 38.7 t/s with server-side acceptance **0/2034** (vs 125.1 t/s + 41/87 first load). Same image, flags, model, boot; only a container restart between loads.

So acceptance looks **load-dependent (bimodal ~0.5 vs 0)** rather than cleanly knob-gated, at least on my box. The env knob name/semantics check out in the patch (`ggml_cuda_try_fuse`, batch-size-gated — note nothing in the condition actually excludes MTP verify batches despite the “Prefill MMQ path” comment, so the mechanism I described remains plausible for the 0-case).

I'm now mapping the distribution: `nofusion` ×3 same-boot with per-load acceptance + tg, plus a fused control. Will report back with the full table and either confirm, refine, or retract. Sorry for the noise — didn't want you chasing a single-sample headline.

**Comment by laurentlp** [comment]
Correction — retracting the smoking gun. Extended testing tonight falsifies the clean bisect story. Full chronological record, same boot, same image/flags/model, temp-0 deterministic prompts, container restarted between loads (fresh process each):

| # | entry (fusion?) | probe acc | tg | verdict |
|---|---|---|---|---|
| 1 | fused, bench | — | 39.3 | BAD |
| 2 | fused, warm rerun (same load) | — | 39.5 | BAD |
| 3 | fused (bf16kv entry) | — | 39.7 | BAD |
| 4 | fused, probe | 0/242 (0.0) | 38.8 | BAD |
| 5 | nofusion, bench | — | 125.1 | GOOD |
| 6 | nofusion, probe (same load) | 41/87 (0.47) | 83.2 | GOOD |
| 7 | nofusion, rerun | 0/2034 srv (0.0) | 38.7 | BAD |
| 8 | nofusion | 42/83 (0.51) | 125.5 | GOOD |
| 9 | nofusion | 42/83 (0.51) | 124.7 | GOOD |
| 10 | nofusion, probe | 42/83 (0.51) | 83.9 | GOOD |
| 11 | fused, probe | 42/83 (0.51) | 83.3 | GOOD |
| 12 | fused, probe (reload) | 42/83 (0.51) | 83.9 | GOOD |

(The identical 42/83 whenever healthy confirms drafts are deterministic — the flip is in verify, not sampling.)

So **both arms flip between 0.0 and ~0.5 across identical loads**, and the late window (8–12) is healthy on *both* settings — including fused twice. Counts (fused 4bad/2good, nofusion 1bad/4good) are suggestive but not significant, and the late fused-healthy runs contradict any deterministic gate story. The knob's effect is unproven; the real finding is **load/time-dependent MTP acceptance bimodality** on this box.

What varies across identical loads here: ASLR addresses, thread scheduling, GPU clocks/thermals over the evening, VRAM reuse contents. Candidates in this area: an address/shape-sensitive dispatch branch, an uninitialized read in the verify path, or a timing-sensitive sync gap (the block does add async staging copies). I can't distinguish these from outside — but the 42/83 determinism-when-healthy plus the early-bad/late-good drift may mean something to you.

Sorry again for the whiplash — single-sample headlines, lesson learned. The box + runner stay available if you want specific probes run.
